apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/UrlSigner.cs

```diff
@@ -60,21 +62,124 @@ private UrlSigner(IBlobSigner blobSigner, IClock clock)
+        /// <remarks>
+        /// A request to the IAM API is executed for signing which increases latency
+        /// as compared with <see cref="FromCredential(ServiceAccountCredential)"/> 
+        /// and that counts towards IAM API quoata consumption.
+        /// </remarks>
+        public static UrlSigner FromCredential(ImpersonatedCredential credential) =>
+            FromBlobSigner(new CredentialBlobSigner(GaxPreconditions.CheckNotNull(credential, nameof(credential)), credential.TargetPrincipal, CredentialsDefaultAlgorithm));
+
+        /// <summary>
+        /// Creates a new <see cref="UrlSigner"/> instance for a compute credential.
+        /// </summary>
+        /// <param name="credential">The compute credential. Must not be null.</param>
+        /// <param name="cancellationToken">A token to cancel the task.</param>
+        /// <remarks>
+        /// <para>
+        /// A request to the IAM API is executed for signing which increases latency
+        /// as compared with <see cref="FromCredential(ServiceAccountCredential)"/> 
+        /// and that counts towards IAM API quoata consumption.
+        /// </para>
+        /// <para>
+        /// The first time a compute credential is used for signing, a request to the
+        /// metadata server is executed, to obtain the associated service account ID,
+        /// which may increase latency of the overall signing request.
+        /// </para>
+        /// </remarks>
+        public static async Task<UrlSigner> FromCredentialAsync(ComputeCredential credential, CancellationToken cancellationToken = default) =>
+            FromBlobSigner(new CredentialBlobSigner(
+                GaxPreconditions.CheckNotNull(credential, nameof(credential)),
+                await credential.GetDefaultServiceAccountEmailAsync(cancellationToken).ConfigureAwait(false),
+                CredentialsDefaultAlgorithm));
+
+        /// <summary>
+        /// Creates a new <see cref="UrlSigner"/> instance for a Google credential if
+        /// the wrapped credential type is supported for signing. In that case, this method
+        /// relies on the appropiate UrlSigner.FromCredential overload.
+        /// </summary>
+        /// <param name="credential">
+        /// The Google credential. Must not be null.
+        /// Must wrap a credential currently supported for signing.
+        /// </param>
+        /// <param name="cancellationToken">The cancellation token to cancel the operation.</param>
+        /// <exception cref="InvalidOperationException">
+        /// The credential wrapped by the Google credential is not currently supported for signing.
+        /// </exception>
+        /// <remarks>
+        /// Please see the specific UrlSigner.FromCredential overloads for more information.
+        /// </remarks>
+        public static async Task<UrlSigner> FromCredentialAsync(GoogleCredential credential, CancellationToken cancellationToken = default) =>
```

Comment:
Does this definitely need to be asynchronous? Can't we wait until the blob signing operation (which is already asynchronous) to get the target principal?